src/qonnx/transformation/quantize_graph.py

```diff
@@ -0,0 +1,261 @@
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of qonnx nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+import numpy as np
+import onnx
+from onnx import TensorProto
+
+from qonnx.core.datatype import DataType
+from qonnx.core.modelwrapper import ModelWrapper
+from qonnx.transformation.base import Transformation
+from qonnx.transformation.general import SortGraph
+from qonnx.transformation.infer_shapes import InferShapes
+from qonnx.util.basic import qonnx_make_model
+from qonnx.util.cleanup import cleanup_model
+
+
+class graph_util:
+    def get_node_id(self, model):
+        node_index = {}
+        node_ind = 0
+        for node in model.graph.node:
+            node_index[node.name] = node_ind
+            node_ind += 1
+        return node_index
+
+    def node_from_name(self, model, node_name):
+        for node in model.graph.node:
+            if node.name == node_name:
+                return node
+
+    def identify_nodes(self, model, node_type):
+        node_list = []
+        for node in model.graph.node:
+            if node.op_type == node_type:
+                node_list.append(node)
+        return node_list
+
+    def create_node(
+        self,
+        model,
+        quantnode_input,
+        quantnode_output_shape,
+        node_count,
+        tensor_count,
+        scale_value,
+        zeropoint_value,
+        bitwidth_value,
+        narrow,
+        signed,
+        rounding_mode,
+    ):
+        quantnode_output_dtype = DataType["UINT8"]
+        quant_tensor = onnx.helper.make_tensor_value_info(
+            model.make_new_valueinfo_name(), TensorProto.FLOAT, quantnode_output_shape
+        )
+        model.graph.value_info.append(quant_tensor)
+        model.set_tensor_datatype(quant_tensor.name, quantnode_output_dtype)
```

Comment:
it is not always guaranteed that the output datatype annotation for a Quant node will be UINT8. sometimes it will be INT8, and other times it will be a SCALED_INT8. better to remove this annotation entirely and let InferDataTypes deal with it prior to inference cost calculation.

Reply:
suggested changes were made in the latest update.